File: say-numbers-js/say-numbers.js
/*
Transform integer numbers to their english word!
*/

export default function getWord(number) {


  const numbStr = String(number);
  const numbLen = numbStr.length;
  console.log("\n-----\nstartnumber", numbStr, "\n----");

  const dict = [];

  dict[0] = new Map([
    [0, "zero"],
    [1, "one"],
    [2, "two"],
    [3, "three"],
    [4, "four"],
    [5, "five"],
    [6, "six"],
    [7, "seven"],
    [8, "eight"],
    [9, "nine"],
    [1_0, "ten"],
    [1_1, "eleven"],
    [1_2, "twelve"],
    [1_3, "thirteen"],
    [1_4, "fourteen"],
    [1_5, "fifteen"],
    [1_6, "sixteen"],
    [1_7, "seventeen"],
    [1_8, "eighteen"],
    [1_9, "nineteen"],
  ]);

  dict[1] = new Map([
    [2, "twenty"],
    [3, "thirty"],
    [4, "fourty"],
    [5, "fifty"],
    [6, "sixty"],
    [7, "fifty"],
    [8, "eighty"],
    [9, "ninety"],
  ]);

  dict[2] = new Map([
    [0, "hundred"],
    [1, "thousand"],
    [2, "million"],
    [3, "billion"],
  ])


const splitIntoArray = () => {
  let j = 0;
  let arr = [];
  let arrDepth = 0;
  for (let i = numbLen; i > 0; i--) {
    if ((j >= 3)) {
      j = 0;
      arrDepth++;
    } 
    j++;
    if (!arr[arrDepth]) {
      arr[arrDepth] = [];
    };
    arr[arrDepth].push(numbStr[i-1])

    //pad the array with 0
    if (i === 1 && arr[arrDepth].length < 3) {
      const pad = new Array(3 - arr[arrDepth].length).fill("0")
      arr[arrDepth] = [...arr[arrDepth], ...pad]
    }
  }

  return arr;

}


const getStringFromSplitArray = (splitArr) => {

  console.log("shplitarr", 
    splitArr, 
    splitArr.length === 1,
    splitArr[0].every(item => item == 0)
  )

  const strArr = [];

  for (let block = splitArr.length - 1; block > -1; block--) {

    console.log("blocki", splitArr[block])
    if (!strArr[block]) {
      strArr[block] = [];
    };


    //speical case for zero
    if (splitArr.length === 1 && splitArr[0].every(item => item == 0)) {
      strArr[block].push(dict[0].get(0))
      return strArr
    }

    const currBlock = splitArr[block];
    strArr[block].push(getWordOfBlock(splitArr[block]).join(" "))
    console.log("block0word", getWordOfBlock(splitArr[block]))

    if (block !== 0) {
      console.log("yo", dict[2].get( parseInt( block )))
      strArr[block].push(dict[2].get( parseInt( block )))
    }
    strArr[block] = strArr[block].join(" ")

    //    for (let numIndex = 0; numIndex < splitArr[block].length; numIndex++) {
/*     for (let numIndex = splitArr[block].length -1 ; numIndex > -1; numIndex--) {
      console.log("num", numIndex, splitArr[block][numIndex])

      const currNumber = splitArr[block][numIndex];
      const prevNumber = splitArr[block][numIndex-1];

      if (numIndex !== 2 && currNumber == 0) {
        continue;
      }

      if (numIndex === 2 &&  splitArr[block][2] != 0) {
        let suffix = dict[2].get( parseInt( 0 ));
        let currStr = dict[0].get( parseInt(currNumber  ));
        console.log("num in hun", currNumber, "currBlock", block, suffix);
        strArr[block].push(currStr, suffix);

        continue;

      }

      if (numIndex === 1) {
        console.log("second pos", currNumber+prevNumber)
        const specNum = dict[0].get( parseInt(currNumber+prevNumber))
        if (specNum !== undefined) {
          console.log("sepcnum", specNum)
          strArr[block].push(specNum);
          break;
        }
        const tysStr = dict[1].get(parseInt(currNumber));
        const abc = dict[0].get( parseInt(prevNumber));
        const tensStr = (abc == "zero") ? "" : "-"+abc
        console.log("tys", tysStr, "tens", abc);
        strArr[block].push(`${tysStr}${tensStr}`);
        break;
        continue;
      }

      let str = dict[0].get( parseInt(currNumber  ));
      console.log("yo")
      strArr[block].push(str)
    } */



  }
  console.log(strArr.reverse().join(" "))
  return strArr.reverse().join(" ")

}

const split = splitIntoArray();
console.log("splitt", split)
const stringi = getStringFromSplitArray(split);

console.log("\n-----\nstartnumber", numbStr, "\nstringi", stringi);

function getWordOfBlock(block) {
  const blockResult = []
  if (block[2] != 0) {
    console.log("block2", block[2])
    blockResult.push(dict[0].get(parseInt(block[2])), dict[2].get(0))
  }

  console.log("getbloy", block[1]+block[0], parseInt(block[1]+block[0]))
  if (block[1] == "1") {

    if (dict[0].get(parseInt(block[1]+block[0])) !== undefined) {
      blockResult.push(dict[0].get(parseInt(block[1]+block[0])));
      return blockResult;
    }
  }

  if (block[1] !== "0") {
    const rest = `${dict[1].get(parseInt(block[1]))}-${dict[0].get(parseInt(block[0]))}`
    blockResult.push(rest)
    return blockResult
  }

  if (block[0] !== "0") {
    blockResult.push(dict[0].get(parseInt(block[0])))
    return blockResult
  }

  return blockResult

}



/**
 * split
 * getStringsFromArray
 * printString
 */

}
/*
getWord(513)
getWord(41)
getWord(234)
getWord(19)
getWord(1_519)
getWord(102_519)
getWord(123_519)
getWord(13_519)

getWord(51_123_519)
getWord(523_123_519)
getWord(10_516_123_519)
getWord(1_100_516_123_519)
getWord(987_654_321_123)
getWord(222)
*/

getWord(0)
getWord(1)
getWord(12)
getWord(10)
getWord(100)
getWord(5_123_519)


//getWord(52123519)


//getWord(1149)
//getWord(12049)
//getWord(22049)

File: say-numbers/SayNumbers.java
package de.entwicklerheld.sayNumbersJava;
import java.util.ArrayList;

public class SayNumbersJava {
    
    private static final int MINIMUM_NUMBER = 0;
    private static final long MAXIMUM_NUMBER = 999_999_999_999L;

    private static final String[][] dict = { 
        {
            "zero",
            "one", "two", "three", "four", "five",
            "six", "seven", "eight", "nine", "ten",
            "eleven", "twelve", "thirteen", "fourteen", "fifteen", 
            "sixteen", "seventeen", "eighteen", "nineteen",
        },
        {
            "", "",
            "twenty", "thirty", "forty", "fifty",
            "sixty", "fifty", "eighty", "ninety"
        },
        {
            "hundred", "thousand", "million", "billion"
        }
    };


    public static String say(long number) {

        if (number < MINIMUM_NUMBER || number > MAXIMUM_NUMBER)  {
            String errorMessage = String.format(
                "Please provide a number between %d – %d", 
                MINIMUM_NUMBER,
                MAXIMUM_NUMBER
            );
            throw new IllegalArgumentException(errorMessage);
        }

        return chunksToString( numberToChunks(number) );
    }

    /**
     * Create an ArrayList that contains an ArrayList of Integers, split into block of three, from the supplied number.
     * E.g. something like the following [ [1, 2, 3], [1, 0, 0] ] for the number 1123
     * @param number
     * @return ArrayList<ArrayList<Integer>>
     */
    private static ArrayList<ArrayList<Integer>> numberToChunks(long number) {
        String numberStr = Long.toString(number);
        int numberLength = numberStr.length();            
        ArrayList<ArrayList<Integer>> chunks = new ArrayList<>();
        final int MAX_CHUNK_SIZE = 3;

        int currentChunkSize = 0;
        int chunksDepth = 0;

        for (int i = numberLength - 1; i >= 0; i--) {

            if (currentChunkSize == MAX_CHUNK_SIZE) {
                chunksDepth++;
                currentChunkSize = 0;
            }
            currentChunkSize++;

            if (chunksDepth >= chunks.size()) {
                chunks.add(new ArrayList<>());
            }

            ArrayList<Integer> currentChunk = chunks.get(chunksDepth);
            Integer currentNumber = Character.getNumericValue(numberStr.charAt(i));
            currentChunk.add(currentNumber);

            //pad array with '0's, if necessary
            if (i == 0) {
                if (currentChunk.size() < MAX_CHUNK_SIZE) {
                    int paddingCount = MAX_CHUNK_SIZE - currentChunk.size();
    
                    for (int j = 0; j < paddingCount; j++) {
                        currentChunk.add(0);
                    }
                }
            }

        }
        return chunks;
    }

    private static String chunksToString(ArrayList<ArrayList<Integer>> chunks) {
        
        ArrayList<String> strArr = new ArrayList<>();

        for (int chunkDepth = chunks.size() - 1; chunkDepth >= 0; chunkDepth--) {
            
            // special handling for 0 - "zero"
            if (chunks.size() == 1 && isChunkAllZeroes(chunks.get(0))) {
                strArr.add(dict[0][0]);
                break;
            }

            // get String of chunk
            strArr.add(
                String.join(" ", chunkToStringArray(chunks.get(chunkDepth)))
            );

            // add "thousand" / "million" / etc. suffix if required, depeding on on "chunkDepth"
            if (chunkDepth != 0 && !isChunkAllZeroes(chunks.get(chunkDepth))) {
                strArr.add(dict[2][chunkDepth]);
            }

        }

        return String.join(" ", strArr).trim();
    }

    private static ArrayList<String> chunkToStringArray(ArrayList<Integer> chunk) {
        ArrayList<String> result = new ArrayList<>();

        // add "hundred" to string, if required, then continue
        if (chunk.get(2) != 0) {
            result.add(dict[0][chunk.get(2)]);
            result.add(dict[2][0]); //adds "hundred" to string            
        }

        if (chunk.get(1) == 1) {
            int concatNumber = Integer.parseInt( "" + chunk.get(1) + chunk.get(0) );
            Boolean isInBounds = ((concatNumber >= 0) && (concatNumber < dict[0].length));
            if (isInBounds) {
                String concatNumberStr = dict[0][concatNumber];
                result.add(concatNumberStr);
                return result;
            }
        }

        if (chunk.get(1) != 0) {
            String str = dict[1][chunk.get(1)];

            if (chunk.get(0) != 0) {
                str += "-" + dict[0][chunk.get(0)];
            }
            result.add(str);
            return result;
        }

        if (chunk.get(0) != 0) {
            result.add(dict[0][chunk.get(0)]);
            return result;
        }

        return result;
    }

    private static Boolean isChunkAllZeroes(ArrayList<Integer> intChunk) {
        for (int i = 0; i < intChunk.size(); i++) {
          if (intChunk.get(i) != 0) return false;
        }
        return true;
    }

}